Sources/WeaverDI/ThreadSafeBuilderStoringDecorator.swift

```diff
@@ -0,0 +1,47 @@
+//
+//  ThreadSafeBuilderStoringDecorator.swift
+//  WeaverDIPackageDescription
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+
+class ThreadSafeBuilderStoringDecorator: BuilderStoring {
+    
+    let builderStoring: BuilderStoring
+    private let queue: DispatchQueue = DispatchQueue(label: "ThreadSafeBuilderStoringDecorator")
+    
+    init(builderStoring: BuilderStoring) {
+        self.builderStoring = builderStoring
+    }
+    
+    func get<B>(for key: InstanceKey, isCalledFromAChild: Bool) -> B? {
+        let semaphore: DispatchSemaphore = DispatchSemaphore(value: 0)
```

Comment:
This semaphore is not needed. You can write it this way:

```swift
var getValue: B? = nil
        
queue.sync {
      getValue = self.builderStoring.get(for: key, isCalledFromAChild: isCalledFromAChild)
}
        
return getValue
```

Reply:
Ok, it's more elegantly this way!